global.json

```diff
@@ -1,6 +1,6 @@
 {
-    "sdk": {
-        "version": "6.0.200",
-        "rollForward": "latestMajor"
-    }
+  "sdk": {
+    "version": "6.0.404",
+    "rollForward": "disable"
```

Comment:
unless there's a _really_ compelling, specific reason to stay on 6.0.404, I would want this line removed.

Reply:
I detailed this in the PR description. Did you read that?

Reply:
I did - I just vehemently disagree with the position (as the .NET SDK PM :P)

Reply:
I had no idea you were the .NET SDK PM 🙂

Not that it changes my argument, but it's a great learning opportunity for me. So let me be more specific so that you can enlighten me as to why I am probably wrong. (And I can of course remove this line even if we agree to disagree.)

In https://github.com/dotnet/fsharp/issues/13678, the OP experienced builds sudenly breaking on CI. I experienced the same at my workplace, which caused some consternation as it broke all our CI builds and prevented me from publishing updates to our services for some hours while I debugged the issue and fixed the build.

The issue only happens if you follow certain best practices regarding repeatable builds, such as restoring packages with lock files (and maybe other stuff, e.g. added by Dotnet.ReproducibleBuilds). The core of the issue was that the FSharp.Core package shipped with the SDK and the corresponding version of FSharp.Core on NuGet had different content hashes. (As I understand it, this was an error on the F# team's part, the details of which can be read in the thread.) The result was that you could install/restore FSharp.Core locally, but not on build servers, depending on the environment on the dev machines and servers (the .NET SDK version, I guess).

One very simple workaround (that saved my day) was to lock the .NET SDK to the previous (patch) version, which did not have this issue. Hopefully, the issue would then be fixed in the next SDK update. (The issue actually occurred again just after, before they finally fixed it.)

Having been burned by this, I chose to pin the exact SDK version at work, and later also for my OSS projects. My experience was that disabling roll-forward was necessary for repeatable builds, since it broke and caused trouble for us when I didn't. You live, you learn. According to the research I did, pinning the .NET SDK version is something that most people don't do - it's a tradeoff of up-to-date SDKs vs. completely repeatable builds - but since I was actually burned by it, due to factors outside my control, I chose to do it for my projects.

I'm sure there are errors in my reasoning, and I look forward to being corrected and learning more. 🙂 For example, the issue discusses `DisableImplicitLibraryPacksFolder`, which (despite me having participated actively in that discussion) I still  don't fully understand. Reading over it again now, it seems like enabling this would mean that `FSharp.Core` is always restored from NuGet, never the SDK, and therefore fix similar problems without having to pin the SDK version. Is that correct? Since `DisableImplicitLibraryPacksFolder` is set to `true` in this PR, that would mean that pinning (here and in all my other projects, both OSS and at work) is unnecessary.

By the way, why specify a verion in `global.json` at all - or at least, why specify .NET 7, as you requested above? Why not specify the lowest version that will work in `global.json`, and allow major version roll forward?

Reply:
This is an excellent comment, but I will be away for the next couple hours on errands so it'll be a bit before I can give it the response it deserves 😅

Reply:
The first major point you seem to be making is around point-in-time fixes for specific issues. In that one, singular, circumstance it's fine to pin explicitly with a combination of version number + disabling roll forward.  In general, though, we hope to provide ways to workaround specific problems (like the FSharp.Core one you link to here) so that the uninterrupted update of SDK tooling can continue.  For this specific case, the workaround that would allow continued updates would be to disable the implicit FSharp.Core reference and pin to a precise one from NuGet, along with disabling the SDK-default 'implicit library packs' mechanism.

The implicit library packs mechanism is a way for the SDK to ship specific versions of packages as part of the SDK install. The main goal of this is to allow more 'zero restore required' workflows - i.e. you should be able to create and build an F# project without needing to connect to NuGet.org.  This however relies on the packages being shipped in the SDK to match exactly with this from NuGet (as you found).  

Overall, if you disable this mechanism entirely, then you should be able to manage FSharp.Core manually and have it always be restored from NuGet rather than the SDK.

Going forward, the SDK is planning to work with the NuGet team to make the 'library packs' mechanism work with features like Package Source Mapping as well, there's a bit of friction between the mechanisms currently.

Reply:
I see. I have now removed that line. However, I have several questions:
* This defaults `rollForward` to `latestPatch`. If we always want to use the latest .NET SDK, should we set `latestMajor`? If not, why not, and then what about other `rollForward` values?
* If we want `latestMajor`, then why have a `global.json` at all? As I understand it, not having it is the same as `rollForward` set to `latestMajor`, except that when it's present, we require at least the version specified in `global.json`. Unless we need an SDK version that is so recent that we can't expect contributors or build servers to have it, it seems we can do without `global.json`.
* If we want `latestMajor` and despite my reasoning above still want `global.json`, why set it to a recent SDK version and not the earliest possible version that will work?

Reply:
There are I think three main use cases that global.json helps with (and that we (SDK team) should support and document):

1. I want to be on latest all the time - aka delete global.json. This is our stated goal, and anything that prevents users from doing this is A Problem to me.
2. I want to establish a minimum bar for contribution to this repository - aka global.json with a version and rollForward set to `latestMajor`. This is the next-most ideal state IMO.
3. I have a very particular bug/regression that the SDK is causing - aka global.json with version and rollForward set to `disabled`. This is the least ideal state, and should only be a short-term decision, until the next monthly SDK release.

Ideally there would be some way to express case 2 easily, because it does have value - it clearly declares what this repo needs in order to build.

Reply:
I see. Thank you for those clarifications. I am confused, however, since you merged this PR in its current state. I would assume that you either wanted to specify `latestMajor` in `global.json`, or else remove `global.json` entirely. I'm happy to make a PR with one of those changes. Would you like that?

Reply:
Just thought of something: Based on the [Releases and support for .NET](https://learn.microsoft.com/en-us/dotnet/core/releases-and-support) page, there can be breaking changes between major and minor SDK versions. Is it then really a good idea to allow automatic roll-forward using `latestMajor` or `latestMinor`? Couldn't that cause problems when breaking changes occur? Also (asking purely for my own interest now), does the use of self-contained apps impact that decision, since AFAIK then the runtime .NET version is determined by the SDK it's built with?